Use CircleCI 2.0 configuration.

lognaturel opened the pull request:
Closes #1585 by copying all reports
Closes #1609 by putting all gradle tasks together

Overall the config is cleaner and the build is faster than using CircleCI 1 so I'd be in favor of switching over and moving integration tests to a specialized platform as discussed with @yanokwa and @jknightco in Slack. There's still some troubleshooting to do with unit tests, though, and I'd like some other ideas on that. First, I tried to run `./gradlew test` but `testOdkCollectReleaseUnitTest` and `testReleaseUnitTest` fail because of trying to write Timber logs to Firebase. [This article](https://firebase.googleblog.com/2016/12/how-does-firebase-initialize-on-android.html) explains how FirebaseCrash is initialized. I thought I could possibly use `MockContext` to run `FirebaseApp.initializeApp` but had no luck with that. @jknightco I bet you have a better way to structure the code that defines `CrashReportingTree` and sets it for release builds so that it doesn't disrupt tests. #1471 also explored some of these issues.

Then I tried running `./gradlew testDebugUnitTest` but those tests don't seem to be running for some reason. I can't think of any reason why that would be the case and I have verified that command works locally. (**EDIT**: got it working)

When it comes to integration tests, it's possible we could get them working here but I didn't try hard since we're strongly considering moving them elsewhere anyway. https://github.com/flyve-mdm/flyve-mdm-android-inventory-agent/blob/develop/.circleci/config.yml seems to be a config that works.

The app's target SDK is 22. I used a 23 Docker image since that's the lowest API level provided by CircleCI - https://hub.docker.com/r/circleci/android/tags/

lognaturel committed:
Use CircleCI 2.0 configuration.

yanokwa commented:
I know this is WIP, but one thing that I'd add is a link to wherever you found the base config you are using.

lognaturel commented:
Running `testDebugUnitTest` on its own didn't work but it looks like running it on the same line as other gradle tasks works. I can't explain that but it is a better solution because it means gradle can make sure redundant tasks aren't run.

I think the only thing that isn't working with this config now is code coverage. I thought the report was supposed to be generated when the tests were run. I tried to run the jacoco debug gradle task but that didn't work because it tries to run the connected tests. @shobhitagarwal1612 any ideas?

yanokwa commented on .circleci/config.yml:
```diff
@@ -0,0 +1,48 @@
+# Reference used: https://circleci.com/docs/2.0/language-android/#docker-images
```
Link to https://circleci.com/docs/2.0/language-android

yanokwa commented on .circleci/config.yml:
```diff
@@ -0,0 +1,48 @@
+    working_directory: ~/code
+
+    docker:
+      - image: circleci/android:api-23-alpha
```
Since there is no reason for this to be a particular version, I'd use the latest version. It looks like a bug if it's 23 and the target SDK is 22.

yanokwa commented on .circleci/config.yml:
```diff
@@ -0,0 +1,48 @@
+      - image: circleci/android:api-23-alpha
+
+    environment:
+      # https://discuss.circleci.com/t/circle-ci-v2-and-android-memory-issues/11207/9
```
Since the link has a bunch of things to try, I'd add a comment that says this is the min change that fixed the problem.

lognaturel commented:
@yanokwa Thanks, I have addressed your comments! 

I just realized that code coverage is currently taken care of by the `jacoco-everywhere` plugin which requires that the gradle `connectedCheck` task be run. Once we finalize the decision of where connected checks should run, we can revisit if and how to get code coverage stats.
